pydm/widgets/rules.py

```diff
@@ -0,0 +1,155 @@
+
+        # Buffer for data coming from the channels
+        self.channels_connection = [False] * len(self.channels_list)
+        self.channels_value = [None] * len(self.channels_list)
+
+        self.channels = []
+
+        for idx, ch in enumerate(self.channels_list):
+            partial_connection = functools.partial(self.channel_conn_callback,
+                                                   idx, ch['channel'],
+                                                   ch['trigger'])
+            partial_value = functools.partial(self.channel_value_callback,
+                                              idx, ch['channel'], ch['trigger'])
+            c = PyDMChannel(ch['channel'],
+                            connection_slot=partial_connection,
+                            value_slot=partial_value)
+            if is_pydm_app():
+                self.app.add_connection(c)
+            self.channels.append(c)
+
+    def channel_conn_callback(self, index, channel_name, trigger, value):
+        """
+        Callback executed when a channel connection status is changed.
+
+        Parameters
+        ----------
+        index : int
+            The channel index on the list for this rule.
+        channel_name : str
+            The channel address.
+        trigger : bool
+            Whether or not this channel should trigger a calculation of the
+            expression
+        value : bool
+            Whether or not this channel is connected.
+
+        Returns
+        -------
+        None
+        """
+        self.channels_connection[index] = value
+
+    def channel_value_callback(self, index, channel_name, trigger, value):
+        """
+        Callback executed when a channel receives a new value.
+
+        Parameters
+        ----------
+        index : int
+            The channel index on the list for this rule.
+        channel_name : str
+            The channel address.
+        trigger : bool
+            Whether or not this channel should trigger a calculation of the
+            expression
+        value : any
+            The new value for this channel.
+
+        Returns
+        -------
+        None
+        """
+        self.channels_value[index] = value
+        if trigger:
+            if not all(self.channels_connection):
+                logger.error(
+                    "Rule %s: Not all channels are connected, skipping execution.",
+                    self.name)
+                return
+            self.should_calculate = True
+
+    def run(self):
+        """
+        Main loop of the RulesEngine which runs at 30Hz until a interruption is
+        requested and calculates the expression if a new value is available at
+        one of the trigger channels.
+
+        Returns
+        -------
+        None
+        """
+        while not self.isInterruptionRequested():
+            if self.should_calculate:
+                self.calculate_expression()
+            QThread.msleep(33)  # 30Hz
```

Comment:
A little confused on why we have to sleep. Is this just to rate limit the engine?

Reply:
If that is the case I might swap it to:

```python
if self.should_calculate:
     self.calculate_experession()
     QThread.msleep(33)
```
This way you are always spinning then when you see an update you sleep to make sure not to spam.

Reply:
Honestly after I wrote this as well. If you do what I suggested above I would increase the sleep to a second. I can't imagine a single widget I want bouncing around or hiding/showing at 30 Hz 😨

Reply:
That is the rate limit for the engine, yes!
If we are always spinning we would have a thread in the worst case consuming 100% of the processing if the channel is disconnected or if it never updates... the current implementation we check every 33ms which helps in this case...